benchmarks/replay/replay.cpp

```diff
@@ -0,0 +1,171 @@
+        // Convert "allocate" or "free" string into `action` enum
+        action a =
+            (thrust::get<0>(t) == "allocate") ? action::ALLOCATE : action::FREE;
+        std::size_t size = thrust::get<1>(t);
+
+        // Convert pointer string into an integer
+        uintptr_t p = std::stoll(thrust::get<2>(t), nullptr, 16);
+        return event{a, size, p};
+      });
+
+  return events;
+}
+
+/**
+ * @brief Represents an allocation made during the replay
+ *
+ */
+struct allocation {
+  allocation() = default;
+  allocation(void* p_, std::size_t size_) : p{p_}, size{size_} {}
+  void* p{};
+  std::size_t size{};
+};
+
+/**
+ * @brief Function object for running a replay benchmark with the specified
+ * `device_memory_resource`.
+ *
+ * @tparam MR The type of the `device_memory_resource` to use for allocation
+ * replay
+ */
+template <typename MR>
+struct replay_benchmark {
+  std::unique_ptr<MR> mr_{};
+  std::vector<event> const& events_{};
+
+  /**
+   * @brief Construct a `replay_benchmark` from a list of events and
+   * set of arguments forwarded to the MR constructor.
+   *
+   * @param events The set of allocation events to replay
+   * @param args Variable number of arguments forward to the constructor of MR
+   */
+  template <typename... Args>
+  replay_benchmark(std::vector<event> const& events, Args&&... args)
+      : mr_{new MR{std::forward<Args>(args)...}}, events_{events} {}
+
+  void operator()(benchmark::State& state) {
+    // Maps a pointer from the event log to an active allocation
+    std::unordered_map<uintptr_t, allocation> allocation_map(events_.size());
+
+    for (auto _ : state) {
+      std::for_each(events_.begin(), events_.end(),
+                    [&allocation_map, &state, this](event e) {
+                      if (action::ALLOCATE == e.act) {
+                        auto p = mr_->allocate(e.size);
+                        allocation_map[e.pointer] = allocation{p, e.size};
+                      } else {
+                        auto a = allocation_map[e.pointer];
+                        mr_->deallocate(a.p, e.size);
+                      }
+                    });
+    }
+  }
+};
+
+
+// Usage: REPLAY_BENCHMARK -f "path/to/log/file"
+int main(int argc, char** argv) {
+  // benchmark::Initialize will remove GBench command line arguments it
+  // recognizes and leave any remaining arguments
+  ::benchmark::Initialize(&argc, argv);
+
+  // Parse for replay arguments:
+  cxxopts::Options options(
+      "RMM Replay Benchmark",
+      "Replays and benchmarks allocation activity captured from RMM logging.");
+
+  options.add_options()("f,file", "Name of RMM log file.",
+                        cxxopts::value<std::string>());
+
+  auto result = options.parse(argc, argv);
+
+  // Parse the log file
+  if (result.count("file")) {
+    auto filename = result["file"].as<std::string>();
+    auto events = parse_csv(filename);
+
+    benchmark::RegisterBenchmark(
```

Comment:
I think we'll want to make the resource selectable.

Reply:
I agree and I think this would make good future work. The goal here was to just get the benchmark working. Once the other experimental resources are merged we can do another pass at how to select which resource is used.